src/PMGDQueryHandler.cc

```diff
@@ -365,6 +368,51 @@ int PMGDQueryHandler::add_edge(const protobufs::AddEdge &ce,
     return 0;
 }
 
+int PMGDQueryHandler::update_edge(const protobufs::UpdateEdge &ue,
+                                  PMGDCmdResponse *response)
+{
+    long id = ue.identifier();
+    bool query = ue.has_query_edge();
+
+    // Make sure there is either an id or a query node
+    if (id < 0 && !query) {
+        set_response(response, PMGDCmdResponse::Error, "No way to find edges to be updated\n");
+        return -1;
+    }
+
+    auto it = _cached_edges.find(id);
+    if (it == _cached_edges.end()) {
+        if (!query) {
+            set_response(response, PMGDCmdResponse::Error, "Undefined _ref value used in update\n");
+            return -1;
+        }
+        else {
+            query_edge(ue.query_edge(), response);
+            if (response->error_code() != PMGDCmdResponse::Success)
+                return -1;
+            it = _cached_edges.find(ue.query_edge().identifier());  // id could have been -ve
+        }
+    }
+
+    auto eit = it->second;
+    long updated = 0;
+    for ( ; *eit; eit->next()) {
+        Edge &e = **eit;
+        updated++;
+        for (int i = 0; i < ue.properties_size(); ++i) {
+            const protobufs::Property &p = ue.properties(i);
+            set_property(e, p);
+        }
+        for (int i = 0; i < ue.remove_props_size(); ++i)
+            e.remove_property(ue.remove_props(i).c_str());
```

Comment:
If many nodes/edges are being updated, it would be advantageous to get the StringIDs for the properties in advance instead of converting each property name to a StringID every time it is used.